packages/template-retail-react-app/app/components/island/index.jsx

```diff
@@ -0,0 +1,222 @@
+            (!context || context?.hydrated) &&
+            ref.current &&
+            !ref.current?.childElementCount &&
+            !clientOnly
+        ) {
+            // No actual SSR Content or rendering after CSR navigation --> hydrate
+            hydrate()
+        }
+    })
+
+    useEffect(() => {
+        if (
+            hydrated ||
+            hydrateOn === 'off' ||
+            (context && !context?.hydrated) ||
+            !ref.current ||
+            (!ref.current?.firstElementChild && !ssr && !clientOnly)
+        ) {
+            return
+        }
+
+        if (hydrateOn === 'idle') {
+            if ('requestIdleCallback' in globalThis) {
+                const callbackId = requestIdleCallback(
+                    () => hydrate(),
+                    typeof options === 'object' && options !== null ? options : undefined
+                )
+                return () => cancelIdleCallback(callbackId)
+            }
+            const timeout = typeof options?.timeout === 'number' ? options?.timeout : 200
+            const timeoutId = setTimeout(() => hydrate(), timeout) // Fallback
+            return () => clearTimeout(timeoutId)
+        } else if (hydrateOn === 'visible' && 'IntersectionObserver' in globalThis) {
+            let element = ref.current?.firstElementChild
+            if (!element) {
+                // Special case where `clientOnly` is `true`. For this case the box suppression behavior of
+                // `display="contents"` is problematic. The workaround is to apply the `IntersectionObserver`
+                // to the so far empty container directly. But for this we need to change its `display`
+                // settings.
+                element = ref.current
+                element.style.display = 'inline-block'
+            }
+            const observer = new IntersectionObserver(
+                (entries) => {
+                    for (const entry of entries) {
+                        if (entry.isIntersecting || entry.intersectionRatio > 0) {
+                            observer.unobserve(element)
+                            if (element === ref.current) {
+                                // Special case where `clientOnly` is `true`. Here we need to reset the `display`
+                                // settings we just had to change.
+                                element.style.display = 'contents'
+                            }
+                            hydrate()
+                            break // break on first match
+                        }
+                    }
+                },
+                typeof options === 'object' && options !== null
+                    ? options
+                    : {
+                          rootMargin: '250px'
+                      }
+            )
+            observer.observe(element)
+            return () => observer.disconnect()
+        }
+
+        // `hydrateOn='load'`
+        hydrate()
+    }, [hydrateOn, hydrated])
```

Comment:
How reactive should island be? Do we care about options, context, ssr, clientOnly?

Reply:
* `ssr` is not reactive, it inly reflects the excution environment
* `options` I indeed missed and added
* `clientOnly` is only relevant on the server where there's no reactivity anyway. So it's intentionally not part of the hook's dependencies.